test/parallel/test-promises-unhandled-rejections.js

```diff
@@ -0,0 +1,591 @@
+        if (now - asyncTestLastCheck > 10000) {
+          return fail(new Error('Async test timeout exceeded'));
+        }
+      }, 10);
+      setTimeout(nextAsyncTest, 10);
+    }
+  };
+
+})();
+
+function setupException(fn) {
+  var listeners = process.listeners('uncaughtException');
+  process.removeAllListeners('uncaughtException');
+  process.on('uncaughtException', fn);
+  return function clean() {
+    process.removeListener('uncaughtException', fn);
+    listeners.forEach(function(listener) {
+      process.on('uncaughtException', listener);
+    });
+  };
+}
+
+function clean() {
+  process.removeAllListeners('unhandledRejection');
+  process.removeAllListeners('rejectionHandled');
+}
+
+function onUnhandledSucceed(done, predicate) {
+  clean();
+  process.on('unhandledRejection', function(reason, promise) {
+    try {
+      predicate(reason, promise);
+    } catch (e) {
+      return done(e);
+    }
+    done();
+  });
+}
+
+function onUnhandledFail(done) {
+  clean();
+  process.on('unhandledRejection', function(reason, promise) {
+    done(new Error('unhandledRejection not supposed to be triggered'));
+  });
+  process.on('rejectionHandled', function() {
+    done(new Error('rejectionHandled not supposed to be triggered'));
+  });
+  setTimeout(function() {
+    done();
+  }, 10);
+}
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+  });
+  Promise.reject(e);
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+  });
+  new Promise(function(_, reject) {
+    reject(e);
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+  });
+  new Promise(function(_, reject) {
+    setImmediate(function() {
+      reject(e);
+    });
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+  });
+  new Promise(function(_, reject) {
+    setTimeout(function() {
+      reject(e);
+    }, 1);
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+  });
+  var _reject;
+  var promise = new Promise(function(_, reject) {
+    _reject = reject;
+  })
+  _reject(e);
+  setImmediate(function() {
+    promise.then(assert.fail, function(){});
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+  });
+  var _reject;
+  var promise = new Promise(function(_, reject) {
+    _reject = reject;
+  });
+  setImmediate(function() {
+    promise.then(assert.fail, function(){});
+  });
+  _reject(e);
+});
+
+asyncTest(function(done) {
```

Comment:
When re-throwing new errors in a promise catch, only the re-thrown error should hit unhandledRejection